hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieAvroDataBlock.java

```diff
@@ -206,6 +213,9 @@ public IndexedRecord next() {
         IndexedRecord record = this.reader.read(null, decoder);
         this.dis.skipBytes(recordLength);
         this.readRecords++;
+        if (this.promotedSchema.isPresent()) {
+          return  HoodieAvroUtils.rewriteRecordWithNewSchema(record, this.promotedSchema.get());
```

Comment:
Is it possible we read the record with the `promotedSchema` directly so that there is no need to rewrite each record ?

Reply:
https://github.com/apache/hudi/blob/908ae8d3f54adcd9ba24205c9d66add546079c58/hudi-common/src/main/java/org/apache/hudi/avro/AvroCastingGenericRecord.java this was an implementation I had but I don't think it will do any better performance wise. Since types are wrapped in objects, I don't think rewriting the record will have that much overhead. Am I right about that?

Reply:
Sorry misunderstood the question. Avro doesn't support type promotion from number to string: https://avro.apache.org/docs/1.10.2/spec.html. So this only occurs in that situation.